### ISSUE TITLE
Switching rapidly between audio sources leads to `PlatformException`

### ISSUE THREAD

**Issue by rubenferreira97** [ryanheise/just_audio#1078]
**Which API doesn't behave as documented, and how does it misbehave?**
Calling `.setAudioSource` with `preload: false` throws `PlatformException`.
However in the docs only states:
```
 /// When [preload] is `true`, this method may throw:
  ///
  /// * [Exception] if no audio source has been previously set.
  /// * [PlayerException] if the audio source was unable to be loaded.
  /// * [PlayerInterruptedException] if another audio source was loaded before
  /// this call completed or the player was stopped or disposed of before the
  /// call completed.
```
**Minimal reproduction project**
https://github.com/rubenferreira97/just_audio_bug

**To Reproduce (i.e. user steps, not code)**
1. Run the project
2. Switch fast between the two audios.
3. See error

**Error messages**

```console
E/flutter (11925): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: PlatformException(abort, Loading interrupted, null, null)
E/flutter (11925): #0      AudioPlayer._setPlatformActive.checkInterruption (package:just_audio/just_audio.dart:1247:7)
E/flutter (11925): #1      AudioPlayer._setPlatformActive.setPlatform (package:just_audio/just_audio.dart:1358:11)
E/flutter (11925): <asynchronous suspension>
E/flutter (11925): 
I/flutter (11925): Error PlatformException(abort, Loading interrupted, null, null)
```

```console
E/flutter (11925): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: PlatformException(Platform player 7154b6f3-01e4-4a0c-897d-2be091c3b53a already exists, null, null, null)
E/flutter (11925): #0      StandardMethodCodec.decodeEnvelope (package:flutter/src/services/message_codecs.dart:652:7)
E/flutter (11925): #1      MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:310:18)
E/flutter (11925): <asynchronous suspension>
E/flutter (11925): #2      MethodChannelJustAudio.init (package:just_audio_platform_interface/method_channel_just_audio.dart:13:5)
E/flutter (11925): <asynchronous suspension>
E/flutter (11925): #3      AudioPlayer._setPlatformActive.setPlatform (package:just_audio/just_audio.dart:1341:13)
E/flutter (11925): <asynchronous suspension>
```

**Expected behavior**
Should switch audio and play it without crashes/exceptions.

**Screenshots**

**Desktop (please complete the following information):**
 - OS: Windows 11
 - Browser: Android

**Smartphone (please complete the following information):**
 - Device: Redmi Note 9S
 - OS: Android 12

**Flutter SDK version**
```
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel stable, 3.13.5, on Microsoft Windows [Version 10.0.22621.2283], locale pt-PT)
[√] Windows Version (Installed version of Windows is version 10 or higher)
[√] Android toolchain - develop for Android devices (Android SDK version 34.0.0-rc2)
[√] Chrome - develop for the web
[X] Visual Studio - develop Windows apps
    X Visual Studio not installed; this is necessary to develop Windows apps.
      Download at https://visualstudio.microsoft.com/downloads/.
      Please install the "Desktop development with C++" workload, including all of its default components
[√] Android Studio (version 2022.3)
[√] VS Code (version 1.82.3)
[√] Connected device (4 available)
[√] Network resources

! Doctor found issues in 1 category.
```

**Additional context**

**Comment by ryanheise** [comment]
> **Which API doesn't behave as documented, and how does it misbehave?** Calling `.stop`, `.setAudioSource` and `.play` too fast may cause issues.

Hi, the idea is to explain how it doesn't behave as "documented" which I'll include below:

```
  /// When [preload] is `true`, this method may throw:
  ///
  /// * [Exception] if no audio source has been previously set.
  /// * [PlayerException] if the audio source was unable to be loaded.
  /// * [PlayerInterruptedException] if another audio source was loaded before
  /// this call completed or the player was stopped or disposed of before the
  /// call completed.
```

I think the bug you might be getting at is that the expected behaviour is point 3 above, but the actual behaviour is an unhandled exception.

**Comment by rubenferreira97** [comment]
I think it is a bug because subsequent calls break completely the player. The players become unresponsive after the first crash with this exception: `PlatformException(Platform player 7154b6f3-01e4-4a0c-897d-2be091c3b53a already exists, null, null, null)`

Did you try the reproducible example?

**Comment by ryanheise** [comment]
I'm just trying to help you fill in the first section of the bug report. The exception is expected and actually right there in the documentation (which I'm saying you need to reference when filling in the first section), and hence the problem is not the exception, but rather that the exception is unhandled.

**Comment by rubenferreira97** [comment]
@ryanheise Sorry to bother, I edited my initial issue.
The exact same exceptions throw when `preload = false` so I guess that's not documented. The player becomes broken afterwards, I still think this is a underlying issue. Even if an exception is thrown I think the player should not be in a inconsistent state.

**Comment by ryanheise** [comment]
The underlying problem "is" that the exception is unhandled, since that means that the exception isn't actually being thrown by the API that purports to throw that exception. if it were thrown, then it would be in the correct state, allowing you to set a new audio source etc.

**Comment by firgia** [comment]
I had the same issue, Any news here?

**Comment by johnareid54** [comment]
@firgia "I had the same issue, Any news here?"
I had the same fault today, spent ages trying to find the cause, as a workaround I used different AudioPlayer();

**Comment by afl-dev** [comment]
same

**Comment by Anshuman115** [comment]
The issue is still present, if anyone has any fix, do share.

The issue is reproducible by rapidly changing the audio sources and trying to play, but the issue is catching the error to fix it.

**Comment by firgia** [comment]
This issue is not fully resolved. 
I have fixed this issue by adding debounce at least 500ms and never see this issue anymore at the moment

Update: 

> After a few months, I still got the bug report from my user. When the Exception is triggered, the audio can't play to play any audio except restarting the app. I have no idea about this, Firebase crashlytics can't report this bug and it's happened randomly.

**Comment by Anshuman115** [comment]
> This issue is not fully resolved. I have fixed this issue by adding debounce at least 500ms and never see this issue anymore at the moment

Can you share the code please ?

**Comment by sajidshaikh2027** [comment]
same issue

**Comment by kingrmb** [comment]
> Can you share the code please ?
@firgia

**Comment by firgia** [comment]
@Anshuman115 @kingrmb 

```import 'dart:async';

import 'package:just_audio/just_audio.dart';

class JustAudioService {
  static final JustAudioService _singleton = JustAudioService._internal();

  factory JustAudioService() => _singleton;

  JustAudioService._internal();

  AudioPlayer audio = AudioPlayer();
  Timer? _debounce;

  Future<bool> play(String source) async {
    Completer<bool> completer = Completer<bool>();
    await stop();

    bool isAlreadyPlay = false;
    if (_debounce == null) {
      isAlreadyPlay = true;

      _playSource(source).then(
        (value) {
          if (!completer.isCompleted) completer.complete(value);
        },
      );
    }

    if (_debounce?.isActive ?? false) _debounce?.cancel();
    _debounce = Timer(const Duration(milliseconds: 500), () async {
      if (!isAlreadyPlay) {
        _playSource(source).then(
          (value) {
            if (!completer.isCompleted) completer.complete(value);
          },
        );
      }
      _debounce = null;
    });

    return completer.future;
  }

  Future<bool> _playSource(String source) async {
    try {
      late AudioSource audioSource;
      if (source.startsWith("http")) {
        audioSource = LockCachingAudioSource(Uri.parse(source));
      } else {
        audioSource = AudioSource.file(source);
      }

      await audio.setAudioSource(audioSource);
      await audio.play();

      return true;
    } catch (e) {
      return false;
    }
  }

  Future<void> stop() => audio.stop();
}
```

**Comment by richanshah** [comment]
any complete solution to this ?

**Comment by richanshah** [comment]
@ryanheise  any new version coming up with this issue fix?

**Comment by iamnabink** [comment]
Any fixes/new release to fix this issue?

**Comment by wayupdev** [comment]
Hi @ryanheise any news about this ?

**Comment by dawid-niedzwiecki** [comment]
Please, this makes the package unusable!

**Comment by N0tExpectErr0r** [comment]
> [@Anshuman115](https://github.com/Anshuman115) [@kingrmb](https://github.com/kingrmb)
> 
> ```
> 
> import 'package:just_audio/just_audio.dart';
> 
> class JustAudioService {
>   static final JustAudioService _singleton = JustAudioService._internal();
> 
>   factory JustAudioService() => _singleton;
> 
>   JustAudioService._internal();
> 
>   AudioPlayer audio = AudioPlayer();
>   Timer? _debounce;
> 
>   Future<bool> play(String source) async {
>     Completer<bool> completer = Completer<bool>();
>     await stop();
> 
>     bool isAlreadyPlay = false;
>     if (_debounce == null) {
>       isAlreadyPlay = true;
> 
>       _playSource(source).then(
>         (value) {
>           if (!completer.isCompleted) completer.complete(value);
>         },
>       );
>     }
> 
>     if (_debounce?.isActive ?? false) _debounce?.cancel();
>     _debounce = Timer(const Duration(milliseconds: 500), () async {
>       if (!isAlreadyPlay) {
>         _playSource(source).then(
>           (value) {
>             if (!completer.isCompleted) completer.complete(value);
>           },
>         );
>       }
>       _debounce = null;
>     });
> 
>     return completer.future;
>   }
> 
>   Future<bool> _playSource(String source) async {
>     try {
>       late AudioSource audioSource;
>       if (source.startsWith("http")) {
>         audioSource = LockCachingAudioSource(Uri.parse(source));
>       } else {
>         audioSource = AudioSource.file(source);
>       }
> 
>       await audio.setAudioSource(audioSource);
>       await audio.play();
> 
>       return true;
>     } catch (e) {
>       return false;
>     }
>   }
> 
>   Future<void> stop() => audio.stop();
> }
> ```

I think this may introduce some latency, and crashes might still occur when network requests exceed 500ms.

**Comment by ryanheise** [comment]
The supplied minimal reproduction project does not seem to use `preload: false`. If anyone would like to share a minimal reproduction project (by which I mean, fork this repo and edit the example with minimal changes to reproduce the bug), that may help.

**Comment by cyberpwnn** [comment]
God this plugin is garbage. Cant do just about anything without babying it, littering everything with try/catches and still, it just fails in 5 uses anyways. How am I supposed to handle user interaction which is unpredictable in a plugin that only wants to work in one way, and when it doesnt work once, it doesnt just fail, everything fails, the entire app has to be restarted, no more audio. 

Can you just... stop asking for reproductions and picking apart random inaccuracies and just use your own plugin? Just try to use it, its a god awful experience and should never be used in production. 

For anyone looking for alternatives, I'd suggest looking at media_kit_libs_audio and audioplayers. I abandoned audioplayers 2 years ago but it looks like its a bit more stable now.

**Comment by ryanheise** [comment]
@cyberpwnn ,

The reason why I haven't prioritised this issue (over others that I am currently working hard at), is that the underlying exception was originally intended to indicate an unintended use of the API. That is to say, you are not supposed to interrupt the loading of an audio source. Instead, you should await the first load before starting the second load. It is indeed a bug that the exception to indicate that a load was interrupted is unhandled, but given that it is best for the app to not actually trigger the exception in the first place, I have considered it a low priority (although I might take a look over the next week).

In that light, the reproduction project could be modified to check before interrupting a load in progress:

```dart
if (player.processingState == ProcessingState.loading) {
  // deal with this state as preferred.
  // e.g. surface error to user, or ignore, or wait:
  // await player.processingState.firstWhere((state) => state != ProcessingState.loading);
}
await player.setAsset(audio);
```

If the app rather wants to make the second load wait until the first load is completed where the two loads are triggered from different event handlers, you could use [`package:synchronized`](https://pub.dev/packages/synchronized) to make these two independent blocks of code wait for each other:

```dart
await lock.synchronized(() async {
  await player.setAsset(audio);
});
```

If a someone wants to contribute on this issue so that the exception is surfaced rather than unhandled, I would hint at looking at the `setPlatformActive` method, the returned Future being where this exception is intended to be surfaced. If not, I may take a look in the next few days.

**Comment by ryanheise** [comment]
The latest release should surface the `PlayerInterruptedException` in this scenario.

**Comment by Abu-Abdullah** [comment]
facing this issue some times, i was able to solve it by changing many things like gradle version, NDK version, com.android.application version .... it depends on the android studio that I'm using (I'm using canary channel) and how it is compatible with all the above. not sure if this is really related, but it solves such errors in some cases.

**Comment by ryanheise** [comment]
@Abu-Abdullah , this exception is not related to things like the gradle version, the NDK version or the application version or the Android Studio version. So if you are relying on a particular choice of those versions to make the error go away, I must warn you that this is a completely unreliable approach, because it doesn't take into consideration what this exception actually means. This exception occurs when you try to load one audio source, and then _while_ it's loading, you try to load another audio source. In most cases, what you probably should do is modify your app to wait for the first load to finish before starting the next load, and then you will avoid getting this exception. In cases where you intentionally want to interrupt a load, you may intend to generate and expect to handle this exception, and that is what the latest release allows you to do.

From here, I would suggest to post on StackOverflow if you need help structuring your code, or opening a new issue if you find there is still incorrect behaviour within the plugin.
